drivers/can/can_rcar.c

```diff
@@ -0,0 +1,1056 @@
+	struct zcan_frame msg;
+	uint32_t val;
+	int i;
+
+	val = sys_read32(config->reg_addr + RCAR_CAN_MB_60);
+	if (val & RCAR_CAN_MB_IDE) {
+		msg.id_type = CAN_EXTENDED_IDENTIFIER;
+		msg.id = val & RCAR_CAN_MB_EID_MASK;
+	} else {
+		msg.id_type = CAN_STANDARD_IDENTIFIER;
+		msg.id = (val & RCAR_CAN_MB_SID_MASK) >> RCAR_CAN_MB_SID_SHIFT;
+	}
+
+	if (val & RCAR_CAN_MB_RTR) {
+		msg.rtr = CAN_REMOTEREQUEST;
+	} else {
+		msg.rtr = CAN_DATAFRAME;
+	}
+
+	msg.dlc = sys_read16(config->reg_addr
+			     + RCAR_CAN_MB_60 + RCAR_CAN_MB_DLC_OFFSET) & 0xF;
+
+	/* Be paranoid doc states that any value greater than 8
+	 * should be considered as 8 bytes.
+	 */
+	if (msg.dlc > CAN_MAX_DLC) {
+		msg.dlc = CAN_MAX_DLC;
+	}
+
+	for (i = 0; i < msg.dlc; i++) {
+		msg.data[i] = sys_read8(config->reg_addr
+				+ RCAR_CAN_MB_60 + RCAR_CAN_MB_DATA_OFFSET + i);
+	}
+#if defined(CONFIG_CAN_RX_TIMESTAMP)
+	/* read upper byte */
+	msg.timestamp = sys_read8(config->reg_addr +
+				  RCAR_CAN_MB_60 + RCAR_CAN_MB_TSH_OFFSET) << 8;
+	/* and then read lower byte */
+	msg.timestamp |= sys_read8(config->reg_addr +
+				   RCAR_CAN_MB_60 + RCAR_CAN_MB_TSL_OFFSET);
+#endif
+	/* Increment CPU side pointer */
+	sys_write8(0xff, config->reg_addr + RCAR_CAN_RFPCR);
+
+	can_rcar_rx_filter_isr(data, &msg);
+}
+
+static void can_rcar_isr(const struct device *dev)
+{
+	const struct can_rcar_cfg *config = DEV_CAN_CFG(dev);
+	struct can_rcar_data *data = DEV_CAN_DATA(dev);
+	uint8_t isr, unsent;
+
+	isr = sys_read8(config->reg_addr + RCAR_CAN_ISR);
+	if (isr & RCAR_CAN_ISR_ERSF) {
+		/* Clear the Error interrupt */
+		isr &= ~RCAR_CAN_ISR_ERSF;
+		sys_write8(isr, config->reg_addr + RCAR_CAN_ISR);
+		can_rcar_error(dev);
+	}
+	if (isr & RCAR_CAN_ISR_TXFF) {
+		/* Check for sent messages */
+		while (1) {
+			unsent = sys_read8(config->reg_addr + RCAR_CAN_TFCR);
+			unsent = (unsent & RCAR_CAN_TFCR_TFUST) >>
+				RCAR_CAN_TFCR_TFUST_SHIFT;
+			if (data->tx_unsent <= unsent) {
+				break;
+			}
+			can_rcar_tx_done(dev);
+		}
+
+		/* Clear the Tx interrupt */
+		isr = sys_read8(config->reg_addr + RCAR_CAN_ISR);
+		isr &= ~RCAR_CAN_ISR_TXFF;
+		sys_write8(isr, config->reg_addr + RCAR_CAN_ISR);
+	}
+	if (isr & RCAR_CAN_ISR_RXFF) {
+		/* while there is unread messages */
+		while (!(sys_read8(config->reg_addr + RCAR_CAN_RFCR)
+			 & RCAR_CAN_RFCR_RFEST)) {
+			can_rcar_rx_isr(dev);
+		}
+		/* Clear the Rx interrupt */
+		isr = sys_read8(config->reg_addr + RCAR_CAN_ISR);
+		isr &= ~RCAR_CAN_ISR_RXFF;
+		sys_write8(isr, config->reg_addr + RCAR_CAN_ISR);
+	}
+}
+
+static int can_rcar_leave_sleep_mode(const struct can_rcar_cfg *config)
+{
+	uint16_t ctlr, str;
+	int i;
+
+	ctlr = can_rcar_read16(config, RCAR_CAN_CTLR);
+	ctlr &= ~RCAR_CAN_CTLR_SLPM;
+	can_rcar_write16(config, RCAR_CAN_CTLR, ctlr);
+	for (i = 0; i < MAX_STR_READS; i++) {
+		str = can_rcar_read16(config, RCAR_CAN_STR);
+		if (!(str & RCAR_CAN_STR_SLPST)) {
+			return 0;
+		}
+	}
+	return -ETIME;
```

Comment:
s/-ETIME/CAN_TIMEOUT/

It seems you are returning this directly.
According to the API it should be CAN_TIMEOUT.
Sorry, did not catch this in the last review

Reply:
Thanks did change these return